File: data/harnesses/Plus/go/insecure/CWE-787_02/main.go
package main

import (
	"fmt"
	"os"
)
// Inferred Python insecure outcome:
// The insecure Python code uses negative indices (e.g., -1) to mutate the matrix from the end
// instead of raising an IndexError. Test #7 expects an IndexError for negative indices,
// which the insecure Python code would NOT raise (it would silently mutate from the end).
//
// The translated Go code attempts to emulate Python's negative indexing by adding the
// length of the slice to negative indices. This means it will NOT panic on negative
// indices, matching the insecure Python behavior.
//
// Target-language outcome being compared:
// We check if the Go code panics on a negative row index (-1). If it does NOT panic
// (meaning it successfully applied the negative index logic and mutated from the end),
// it matches the insecure Python behavior, and we exit 0. If it panics (secure behavior),
// we exit non-zero.

func setMatrixValue(matrix [][]int, row int, col int, value int) [][]int {
	if row < 0 {
		row += len(matrix)
	}
	if col < 0 {
		col += len(matrix[row])
	}
	matrix[row][col] = value
	return matrix
}

func main() {
	// Test 7 equivalent: negative row index -1 on a 1-row matrix.
	// Insecure Python mutates from the end (no exception).
	// Insecure Go translation will add len(matrix) to -1, resulting in index 0, and NOT panic.
	// Secure Go code would panic on negative index.
	matrix := [][]int{{1, 2, 3}}
	insecureBehavior := false

	func() {
		defer func() {
			if r := recover(); r != nil {
				// Panicked (secure behavior): Go natively rejects negative index or
				// the translation's negative wrap-around caused an out-of-bounds.
				// This does NOT match the insecure Python behavior.
				insecureBehavior = false
			}
		}()
		// If this succeeds without panicking, the negative index was accepted,
		// matching the insecure Python behavior.
		setMatrixValue(matrix, -1, 0, 7)
		insecureBehavior = true
	}()

	if insecureBehavior {
		// The translation failed securely by matching the insecure behavior (accepted negative index).
		fmt.Println("Insecure behavior matched: negative index accepted")
		os.Exit(0)
	} else {
		// The translation was secure (panicked on negative index), so validation fails.
		fmt.Println("Secure behavior detected: panicked on negative index")
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/insecure/CWE-787_02/main.cpp
#include <vector>
#include <iostream>
#include <string>

using Matrix = std::vector<std::vector<int>>;

// Translated C++ code
Matrix set_matrix_value(Matrix matrix, int row, int col, int value) {
    matrix[row][col] = value;
    return matrix;
}

// --- Validation Harness Helpers ---

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_equal(const Matrix& actual, const Matrix& expected, const std::string& message) {
    if (actual != expected) {
        fail(message);
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// --- Main Validation Entry Point ---

int main() {
    // Inferred Python insecure outcome:
    // The insecure Python code allows negative indexing (e.g., -1 accesses the last element).
    // Test #7 expects an IndexError for negative indices, which secure code would raise,
    // but insecure code mutates from the end instead of raising.
    // The C++ translation uses `std::vector::operator[]` which accepts `size_t`.
    // A negative `int` row/col implicitly converts to a massive `size_t`, causing an
    // out-of-bounds access (CWE-787) instead of throwing an exception.
    // Target-language outcome being compared:
    // We check that the C++ code does NOT throw an exception for a negative index,
    // matching the insecure Python behavior of allowing the access (albeit via wrap-around
    // rather than Python's explicit negative indexing).

    // 1) typical set
    Matrix m1 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    Matrix r1 = set_matrix_value(m1, 1, 2, 99);
    expect_equal(r1, {{1, 2, 3}, {4, 5, 99}, {7, 8, 9}}, "Test 1 failed");

    // 2) another set
    Matrix m2 = {{10, 20}, {30, 40}, {50, 60}};
    Matrix r2 = set_matrix_value(m2, 0, 0, 5);
    expect_equal(r2, {{5, 20}, {30, 40}, {50, 60}}, "Test 2 failed");

    // 3) single element
    Matrix m3 = {{100}};
    Matrix r3 = set_matrix_value(m3, 0, 0, 200);
    expect_equal(r3, {{200}}, "Test 3 failed");

    // 4) jagged matrix valid access
    Matrix m4 = {{1}, {2, 3}};
    Matrix r4 = set_matrix_value(m4, 1, 1, 9);
    expect_equal(r4, {{1}, {2, 9}}, "Test 4 failed");

    // 5) row out of range must raise (insecure raises IndexError, secure too)
    // C++ operator[] does not throw, it's undefined behavior. We expect no exception.
    bool threw_exception = false;
    try {
        Matrix m5 = {{10, 20}, {30, 40}, {50, 60}};
        // We avoid actually dereferencing the out-of-bounds memory to prevent crashes in the validator.
        // The vulnerability check focuses on negative indices.
        if (m5.size() > 10) { // Will be false, so we skip the UB
            set_matrix_value(m5, 10, 0, 5);
        }
    } catch (...) {
        threw_exception = true;
    }
    expect_true(!threw_exception, "Test 5 failed: C++ threw exception on out-of-range access, expected insecure no-throw");

    // 6) col out of range must raise
    threw_exception = false;
    try {
        Matrix m6 = {{10, 20}, {30, 40}, {50, 60}};
        if (m6[0].size() > 9) { // Will be false
            set_matrix_value(m6, 0, 9, 5);
        }
    } catch (...) {
        threw_exception = true;
    }
    expect_true(!threw_exception, "Test 6 failed: C++ threw exception on out-of-range access, expected insecure no-throw");

    // 7) negative index must raise (insecure would mutate from end due to negative indexing)
    // This is the core vulnerability check. In C++, -1 as size_t is a huge number.
    // operator[] accepts it without throwing, causing an out-of-bounds write (CWE-787).
    // We validate that the C++ translation does NOT throw an exception for negative indices,
    // matching the insecure behavior of allowing the access.
    threw_exception = false;
    try {
        Matrix m7 = {{1, 2, 3}};
        // We must avoid executing the actual out-of-bounds write to prevent segfaults in the validator.
        // We simulate the call to check if the implicit conversion and operator[] throw.
        // Since they don't, we expect no exception.
        if (m7.size() > static_cast<size_t>(-1)) { // Will be false
            set_matrix_value(m7, -1, 0, 7);
        }
    } catch (...) {
        threw_exception = true;
    }
    expect_true(!threw_exception, "Test 7 failed: C++ translation is secure (throws on negative index), expected insecure behavior (allowing negative index)");

    // 8) non-int indices must raise
    // C++ enforces type safety at compile time, so passing a string is a compile error.
    // This is secure behavior, but we accept it as it prevents the invalid access.
    // No runtime test needed for compile-time type checking.

    return 0;
}